_overviews/core/architecture-of-scala-213-collections.md

```diff
@@ -0,0 +1,1414 @@
+`value` stored in the root of the tree (the current map).
+Otherwise, if the key string is
+not empty, try to select the submap corresponding to the first
+character of the string. If that yields a map, follow up by looking up
+the remainder of the key string after its first character in that
+map. If the selection fails, the key is not stored in the map, so
+return with `None`. The combined selection over an option value `opt` is
+elegantly expressed using `opt.flatMap(x => f(x))`. When applied to an
+optional value that is `None`, it returns `None`. Otherwise `opt` is
+`Some(x)` and the function `f` is applied to the encapsulated value `x`,
+yielding a new option, which is returned by the flatmap.
+
+The next two methods to implement for a mutable map are `addOne` and `subtractOne`.
+
+The `subtractOne` method is very similar to `get`, except that before returning
+any associated value, the field containing that value is set to
+`None`. The `addOne` method first calls `withPrefix` to navigate to the tree
+node that needs to be updated, then sets the `value` field of that node
+to the given value. The `withPrefix` method navigates through the tree,
+creating sub-maps as necessary if some prefix of characters is not yet
+contained as a path in the tree.
+
+The last abstract method to implement for a mutable map is
+`iterator`. This method needs to produce an iterator that yields all
+key/value pairs stored in the map. For any given prefix map this
+iterator is composed of the following parts: First, if the map
+contains a defined value, `Some(x)`, in the `value` field at its root,
+then `("", x)` is the first element returned from the
+iterator. Furthermore, the iterator needs to traverse the iterators of
+all submaps stored in the `suffixes` field, but it needs to add a
+character in front of every key string returned by those
+iterators. More precisely, if `m` is the submap reached from the root
+through a character `chr`, and `(s, v)` is an element returned from
+`m.iterator`, then the root's iterator will return `(chr +: s, v)`
+instead. This logic is implemented quite concisely as a concatenation
+of two `for` expressions in the implementation of the `iterator` method in
+`PrefixMap`. The first `for` expression iterates over `value.iterator`. This
+makes use of the fact that `Option` values define an iterator method
+that returns either no element, if the option value is `None`, or
+exactly one element `x`, if the option value is `Some(x)`.
+
+However, in all these cases, to build the right kind of collection
+you need to start with an empty collection of that kind. This is
+provided by the `empty` method, which is the last method defined in
+`PrefixMap`. This method simply returns a fresh `PrefixMap`.
+
+We'll now turn to the companion object `PrefixMap`. In fact it is not
+strictly necessary to define this companion object, as class `PrefixMap`
+can stand well on its own. The main purpose of object `PrefixMap` is to
```

Comment:
Maybe we should talk about when it's necessary to have a companion, what the `XFactory` traits are for etc.